integrationtest/backend/elasticsearch/src/test/java/org/hibernate/search/integrationtest/backend/elasticsearch/search/query/ElasticsearchBoolSearchPredicateIT.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.integrationtest.backend.elasticsearch.search.query;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.hibernate.search.engine.backend.document.IndexFieldReference;
+import org.hibernate.search.engine.backend.document.model.dsl.IndexSchemaElement;
+import org.hibernate.search.engine.search.predicate.dsl.SearchPredicateFactory;
+import org.hibernate.search.integrationtest.backend.tck.testsupport.util.rule.SearchSetupHelper;
+import org.hibernate.search.util.impl.integrationtest.mapper.stub.SimpleMappedIndex;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+
+public class ElasticsearchBoolSearchPredicateIT {
+
+	@Rule
+	public final SearchSetupHelper setupHelper = new SearchSetupHelper();
+
+	private final SimpleMappedIndex<IndexBinding> index = SimpleMappedIndex.of( IndexBinding::new );
+
+	@Before
+	public void setup() {
+		setupHelper.start().withIndex( index ).setup();
+	}
+
+	@Test
+	public void resultingQueryOptimization() {
+		SearchPredicateFactory f = index.createScope().predicate();
+
+		assertThat(
+				new Gson().fromJson(
+						index.query()
+								.where( f.bool().must( f.not( f.not( f.not( f.not( f.not( f.not( f.not( f.match().field( "fieldName" ).matching( "test" ) ) ) ) ) ) ) ) ).toPredicate() )
+								.toQuery()
+								.queryString(),
+						JsonObject.class
+				)
+		).isEqualTo( new Gson().fromJson(
```

Comment:
Just so you know, there are utils to compare JSON in a lenient way, e.g.:

https://github.com/hibernate/hibernate-search/blob/fe3b77f29423afcdd6d103b51f1db6316dfd13c9/integrationtest/backend/elasticsearch/src/test/java/org/hibernate/search/integrationtest/backend/elasticsearch/schema/management/ElasticsearchIndexSchemaManagerCreationOrPreservationIT.java#L113-L118

Reply:
nice ! 😃